lib/checks/checks.go

```diff
@@ -319,119 +329,145 @@ func (r *checker) Run(ctx context.Context) error {
 	if ifTestsDisabled() {
 		log.Infof("Skipping single-node checks due to %q set.",
 			constants.PreflightChecksOffEnvVar)
 		return nil
 	}
 
-	var errors []error
-
 	requirements := r.Requirements[server.Server.Role]
 	validateCtx, cancel := context.WithTimeout(ctx, defaults.AgentValidationTimeout)
 	defer cancel()
+
 	failed, err := r.Remote.Validate(validateCtx, server.AdvertiseIP, ValidateConfig{
 		Manifest: r.Manifest,
 		Profile:  server.Server.Role,
 		Docker:   requirements.Docker,
 	})
 	if err != nil {
 		log.WithError(err).Warn("Failed to validate remote node.")
-		errors = append(errors,
-			trace.BadParameter("failed to validate remote node %v", server))
-	}
-	if len(failed) != 0 {
-		errors = append(errors, trace.BadParameter("%v failed checks:\n%v",
-			server, FormatFailedChecks(failed)))
+		failed = append(failed, &agentpb.Probe{
+			Detail: err.Error(),
+			Error:  fmt.Sprintf("failed to validate node %v", server),
+		})
 	}
 
 	err = checkServerProfile(server, requirements)
 	if err != nil {
-		errors = append(errors, err)
+		log.WithError(err).Warn("Failed to validate profile requirements.")
+		failed = append(failed, &agentpb.Probe{
+			Detail: err.Error(),
+			Error:  "failed to validate profile requirements",
+		})
 	}
 
 	dockerConfig := r.Manifest.SystemDocker()
 	if r.TestDockerDevice {
 		err = checkDockerDevice(server, dockerConfig)
 		if err != nil {
-			errors = append(errors, err)
+			log.WithError(err).Warn("Failed to validate docker device.")
+			failed = append(failed, &agentpb.Probe{
+				Detail: err.Error(),
+				Error:  "failed to validate docker device",
+			})
 		}
 	}
 
 	err = checkSystemPackages(server, dockerConfig)
 	if err != nil {
-		errors = append(errors, err)
+		log.WithError(err).Warn("Failed to validate system packages.")
+		failed = append(failed, &agentpb.Probe{
+			Detail: err.Error(),
+			Error:  "failed to validate system packages",
+		})
 	}
 
 	err = r.checkTempDir(ctx, server)
 	if err != nil {
-		errors = append(errors, err)
+		log.WithError(err).Warn("Failed to validate temporary directory.")
+		failed = append(failed, &agentpb.Probe{
+			Detail: err.Error(),
+			Error:  "failed to validate temporary directory",
+		})
 	}
 
 	err = r.checkDisks(ctx, server)
 	if err != nil {
-		errors = append(errors, err)
+		log.WithError(err).Warn("Failed to validate disk requirements.")
+		failed = append(failed, &agentpb.Probe{
+			Detail: err.Error(),
+			Error:  "failed to validate disk requirements",
+		})
 	}
 
-	return trace.NewAggregate(errors...)
+	return failed
 }
 
 // CheckNodes executes checks that take all provided servers into account.
-func (r *checker) CheckNodes(ctx context.Context, servers []Server) error {
+func (r *checker) CheckNodes(ctx context.Context, servers []Server) (failed []*agentpb.Probe) {
 	if ifTestsDisabled() {
 		log.Infof("Skipping multi-node checks due to %q set.",
 			constants.PreflightChecksOffEnvVar)
 		return nil
 	}
 
-	if len(servers) < 2 {
```

Comment:
Turns out this check up here was preventing the `checkPorts` for a single node from executing so bound ports weren't detected.